File: 0130-surrounded-regions/0130-surrounded-regions.java
class Solution {
    int[][] used;
    int[] dr = new int[]{0, 0, 1, -1};
    int[] dc = new int[]{1, -1, 0, 0};
    public void solve(char[][] board) {
        used = new int[board.length][board[0].length];

        for (int i = 1; i < board.length - 1; i++) {
            dfs(board, i, 0);
            dfs(board, i, board[0].length - 1);
        }

        for (int j = 1; j < board[0].length - 1; j++) {
            dfs(board, 0, j);
            dfs(board, board.length - 1, j);
        }

        for (int i = 1; i < board.length - 1; i++) {
            for (int j = 1; j < board[0].length - 1; j++) {
                if (board[i][j] == 'O' && used[i][j] == 0) {
                    board[i][j] = 'X';
                }
            }
        }


    }

    private void dfs(char[][] board, int row, int col) {
        if (row < 0 || row >= board.length || col < 0 || col >= board[0].length || board[row][col] == 'X' || used[row][col] == 1) {
            return;
        }

        used[row][col] = 1;

        for (int k = 0; k < 4; k++) {
            dfs(board, row + dr[k], col + dc[k]);
        }

    }
}

File: 0130-surrounded-regions/0130-surrounded-regions.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        dx, dy = [1, -1, 0, 0], [0, 0, 1, -1]
        m, n = len(board), len(board[0])

        def dfs(x, y):
            board[x][y] = "M"

            for k in range(4):
                nx = x + dx[k]
                ny = y + dy[k]

                if 0 <= nx < m and 0 <= ny < n and board[nx][ny] == "O":
                    dfs(nx, ny)

        for i in range(m):
            if board[i][0] == "O":
                dfs(i, 0)

            if board[i][n - 1] == "O":
                dfs(i, n - 1)

        for j in range(n):
            if board[0][j] == "O":
                dfs(0, j)

            if board[m - 1][j] == "O":
                dfs(m - 1, j)
        
        for i in range(m):
            for j in range(n):
                if board[i][j] == "O":
                    board[i][j] = "X"

                elif board[i][j] == "M":
                    board[i][j] = "O"


